recipes/SignalSnakeJs/src/Main.purs

```diff
@@ -0,0 +1,369 @@
+--
+drawPoint :: Point -> Color -> Context2D -> Effect Unit
+drawPoint { x, y } color ctx = do
+  setFillStyle ctx $ toHexString color
+  fillPath ctx
+    $ rect ctx
+        { x: cellSize * toNumber x
+        , y: cellSize * toNumber y
+        , width: cellSize
+        , height: cellSize
+        }
+
+{-
+Note that we're currently keeping things simple and re-rendering
+the entire canvas from scratch from each state.
+We could be more efficient and just overwrite the cells
+that change, but that increases complexity.
+-}
+render :: Context2D -> Model -> Effect Unit
+render ctx m = do
+  -- Walls
+  setFillStyle ctx $ toHexString wallColor
+  fillPath ctx
+    $ rect ctx
+        { x: 0.0
+        , y: 0.0
+        , width: cellSize * toNumber (xmax + 2)
+        , height: cellSize * toNumber (ymax + 2)
+        }
+  -- Interior
+  setFillStyle ctx $ toHexString bgColor
+  fillPath ctx
+    $ rect ctx
+        { x: cellSize
+        , y: cellSize
+        , width: cellSize * toNumber xmax
+        , height: cellSize * toNumber ymax
+        }
+  -- Snake
+  _ <- for m.snake (\x -> drawPoint x snakeColor ctx)
+  -- Food
+  drawPoint m.food foodColor ctx
+
+-- SIGNALS
+--
+-- An `Action` signal that fires at our tick rate
+sigTicks :: Signal Action
+sigTicks = sampleOn period $ constant Tick
+  where
+  period = every $ second / ticksPerSecond
+
+-- An `Action` signal that fires on arrow key presses.
+-- Note that this signal is wrapped in an Effect,
+-- so requires some unwrapping to work with.
+sigArrowsEff :: Effect (Signal Action)
+sigArrowsEff = do
+  -- Unwrap effects from each keyPressed call
+  left <- keyPressed 37
+  right <- keyPressed 39
+  up <- keyPressed 38
+  down <- keyPressed 40
+  -- This block does a few things (describing from back):
+  -- * Maps each key's Boolean Signal to a Direction with mapKey.
+  -- * Merges all four signals. Note that `<>`/`append` means `merge`.
+  -- * Wraps in a SetDir Action
+  -- * Wraps in an Effect
+  pure $ SetDir
+    <$> mapKey Left left
+    <> mapKey Right right
+    <> mapKey Up up
+    <> mapKey Down down
+
+{- Note that this strategy for merging signals only considers the
+most recent start of a keypress to determine a single key that
+might be pressed.
+So if two keys are pressed between frames (within 17ms @ 60Hz),
+then the first keystroke will be overwritten and ignored.
+A possible solution is to maintain a queue of unhandled
+keystrokes, but this increases complexity.
+-}
+-- Convert a keypress (bool) signal to a Direction signal.
+-- Note that Signals must always have a value, so initialDirection
+-- is used here to provide a signal value at time = 0.
+mapKey :: Direction -> Signal Boolean -> Signal Direction
+mapKey dir sig = filterMap (fromBool dir) initialDirection sig
+
+-- Helper function for mapKey's filterMap.
+-- Convert's a Boolean to a Maybe using a default value.
+fromBool :: forall a. a -> Boolean -> Maybe a
+fromBool x b
+  | b = Just x
+  | otherwise = Nothing
```

Comment:
I suspected something like this would already exist, but I couldn't find anything on Pursuit.

Reply:
I think that Nate showed this on the slack recently:

```purescript
import Control.MonadZero (guard)

fromBool x b = guard b $> x
```